server/middleware.go

```diff
@@ -102,3 +104,60 @@ func withLogging(
 		log.Info("request", args...)
 	}
 }
+
+// withCORS is a middleware that adds CORS headers to responses.
+// It intercepts OPTIONS requests for handling CORS preflight requests.
+func withCORS(
+	handleFn func(http.ResponseWriter, *http.Request),
+	corsConfig config.ServerConfig,
+	log logging.Logger,
+) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// If CORS is not enabled, just call the handler
+		if !corsConfig.CORSEnabled {
+			handleFn(w, r)
+			return
+		}
+
+		origin := r.Header.Get("Origin")
+		if origin == "" {
+			// Not a CORS request
+			handleFn(w, r)
+			return
+		}
+
+		// Check if origin is allowed
+		allowOrigin := "*"
+		if len(corsConfig.CORSAllowedDomains) > 0 {
+			allowed := false
+			for _, domain := range corsConfig.CORSAllowedDomains {
+				if strings.TrimSpace(domain) == origin || 
+				   strings.HasSuffix(origin, strings.TrimSpace(domain)) {
+					allowed = true
+					break
+				}
+			}
+			if !allowed {
+				log.Info("CORS rejected", "origin", origin)
+				handleFn(w, r)
+				return
+			}
+			allowOrigin = origin
+		}
+
+		// Set CORS headers
+		w.Header().Set("Access-Control-Allow-Origin", allowOrigin)
+		w.Header().Set("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
```

Comment:
asked this somewhere else but why is OPTIONS required? might be good to document here since most of us proxy developers are not super familiar with web stuff (at least talking for myself)